### PR TITLE
[feat] Allow to set search arguments inside section class

### PR DESCRIPTION
**Pull request by ineverov**
Issue #85

## Changes
* Allow to provide default search arguments inside section definition
  *  If section has its own search arguments, default search arguments will be ignored
  * If section has no search arguments section class or any parent class deafult arguments will be used
* Allow to provide section class together with a block
* Update README

### CODE REVIEW

**Comment by ineverov** [comment]
going to work on docs/feature tests later this or next week

**Comment by luke-hill** [comment]
Option 1 definitely.

If you provide this argument in the Class, use that.
If you provide the argument in the definition DSL, use that.
If you provide the argument in the definition DSL AND the Class, use the definition DSL (No smart merging)

File: lib/site_prism/section.rb [hunk 1]
```diff
@@ -11,6 +11,30 @@ class Section
 
     attr_reader :root_element, :parent
 
+    class << self
```
Comment (luke-hill) [inline lib/site_prism/section.rb]: This area will probably need some explaining to me, as I'm not quite sure on the moving parts. Would you be willing to jump on slack?

File: spec/section_spec.rb [hunk 2]
```diff
@@ -49,6 +118,38 @@ class PageWithAnonymousSection < SitePrism::Page
   let(:locator) { instance_double('Capybara::Node::Element') }
   let(:section_with_block) { SitePrism::Section.new(Page.new, locator) { 1 + 1 } }
 
+  describe '#default_search_arguments' do
+    let(:base_section) { Class.new(SitePrism::Section) }
```
Comment (luke-hill) [inline spec/section_spec.rb]: I'd prefer a more legible integration-y style mock here. With a physical Page Object being instantiated through a more human interface.

Something like https://github.com/natritmeyer/site_prism/pull/250/files#diff-6fa85b9a8ee926d0d01e58459cddac2cR11 would be a good basis in my eyes

File: spec/section_spec.rb [hunk 3]
```diff
@@ -49,6 +118,38 @@ class PageWithAnonymousSection < SitePrism::Page
...
+        section.set_default_search_arguments '//body', text: 'Body'
+      end
+
+      it 'should return default search arguments' do
+        expect(arguments).to eql(['//body', text: 'Body'])
+      end
+
+      context 'if base class default_search_arguments are set' do
+        before do
+          base_section.set_default_search_arguments :xpath
```
Comment (luke-hill) [inline spec/section_spec.rb]: This seems erroneous, doesn't this just imply that you've set a base searching path of ... nothing? I know you're in a WIP so ignore this if irrelevant

File: lib/site_prism/section.rb (hunk 1 above)
Comment (ineverov) [reply lib/site_prism/section.rb]: yes, sure. Give me a link to it pls

File: spec/section_spec.rb (hunk 2 above)
Comment (ineverov) [reply spec/section_spec.rb]: I did it for the particular reason. Defining class integration-y will affect all unit tests and may potentially lead to false positive results. Because class will remain the same for all the test and redefinition/changes to this class will affect all the tests. For example,
```ruby
context '#hello' do
  class A
     def hello; 'Hello'; end
  end
  it 'should return hello' do
     expect(A.new.hello).to eql('Hello')
  end
end

context '#hi' do
  class A
     def hello; 'hi'; end
  end
  it { expect(A.new.hello).to eql('hi')}
end
```
in this case first example will fail

File: spec/section_spec.rb (hunk 3 above)
Comment (ineverov) [reply spec/section_spec.rb]: the idea was to allow to concatenate search arguments. something like
```ruby
#base class
set_default_search_arguments :xpath, '//html'

#child class
set_default_search_arguments text: 'Name'

# will produce search arguments like: 
# :xpath, '//html', text: 'Name'
```

But I'm going to change it as you recommended in PR comment to always overrride with child arguments if they present

**Comment by ineverov** [comment]
@luke-hill @tgaff Review pls

File: lib/site_prism/section.rb (hunk 1 above)
Comment (luke-hill) [reply lib/site_prism/section.rb]: Happy for this and the below (Although not sure why you can't just use class methods). (Not happy with the memoized default_search_args (2 methods below), but mentioned on slack

File: spec/section_spec.rb (hunk 2 above)
Comment (luke-hill) [reply spec/section_spec.rb]: Correct, due to overloading. However I'm in favour of doing some named examples scoping and having these classes reinstantiated with differing layouts (xpath / css), for example. And have a chunky PR reworking a load of specs for this.

File: lib/site_prism/section.rb (hunk 1 above)
Comment (ineverov) [reply lib/site_prism/section.rb]: It's actually the same as using class methods. I can change it to `def self.` if you prefer this

File: spec/section_spec.rb (hunk 2 above)
Comment (ineverov) [reply spec/section_spec.rb]: Ok, I'll change it

File: lib/site_prism/section.rb (hunk 1 above)
Comment (luke-hill) [reply lib/site_prism/section.rb]: If it doesn't facilitate a big rewrite could you (#newbie)

File: test_site/pages/home.rb [hunk 4]
```diff
@@ -30,7 +30,10 @@ class TestHomePage < SitePrism::Page
   element :nonexistent_element, 'input#nonexistent'
 
   # individual sections
-  section :people, People, '.people'
+  section :people, People
```
[the remark this answers is not in the document]
Comment (ineverov) [reply test_site/pages/home.rb]: It is used in some tests I guess, but yes, I can use below section and rename it to just people

[the remark this answers is not in the document]
Comment (luke-hill) [reply test_site/pages/home.rb]: Actually I've re-read the part above. And you're showing that actually both of these references are the same class object. I'm happy with that.

Could you add a comment to the first item. Saying that the locator has been defined elsewhere and each ref will have both element objects in it

File: spec/sections_spec.rb [hunk 5]
```diff
@@ -22,6 +26,20 @@ class Page < SitePrism::Page
...
+  describe '.section with class and block' do
+    before do
+      expect(subject).to receive(:find_first).and_return(:element)
+    end
+
+    it 'should be an instance of provided section class' do
+      expect(subject.section_with_a_block.class.ancestors).to include(SingleSection)
+    end
+
+    it 'should have elements from the block' do
```
Comment (luke-hill) [inline spec/sections_spec.rb]: Should we therefore add an extra test `it 'has elements from the defined class'`

Then add an extra element definition after opening the class in this file on line8, call it whatever makes you happy. (Just add one)

File: .overcommit.yml [hunk 6]
```diff
@@ -0,0 +1,31 @@
+# Use this file to configure the Overcommit hooks you wish to use. This will
```
Comment (luke-hill) [inline .overcommit.yml]: Is this file meant to be in here>? If it's not, feel free to edit our gitignore

File: README.md [hunk 7]
```diff
@@ -805,10 +805,52 @@ method. It takes 3 arguments: the first is the name of the section as
 referred to on the page (sections that appear on multiple pages can be
 named differently). The second argument is the class of which an
 instance will be created to represent the page section, and the third
-argument is a css selector that identifies the root node of the section
+argument the search argument that identifies the root node of the section
```
Comment (luke-hill) [inline README.md]: This doesn't quite read correctly, probably should be reverted

File: README.md [hunk 8]
```diff
@@ -805,10 +805,52 @@ method. It takes 3 arguments: the first is the name of the section as
...
 on this page (note that the css selector can be different for different
 pages as the whole point of sections is that they can appear in
-different places on different pages).
+different places on different pages). It can also take the block with
```
Comment (luke-hill) [inline README.md]: This definition is defined under Anonymous sections, so isn't required.

File: README.md [hunk 9]
```diff
@@ -805,10 +805,52 @@ method. It takes 3 arguments: the first is the name of the section as
...
+
+```ruby
+class SectionExperiments < SitePrism::Page
+  sections :level_1, '.level-1' do
+    element :element1, '.element-1'
+  end
+end
+```
+
+Providing class and block allows also to extend section for this
```
Comment (luke-hill) [inline README.md]: This can be rewritten to something like ...
```
If you define a section as a class and as an Anonymous section, you can have some handy constructs like the one below

<insert ruby code>
```

I'd also move this advice to below here: https://github.com/natritmeyer/site_prism/pull/250/files#diff-04c6e90faac2675aa89e2176d2eec7d8R1180

File: README.md [hunk 10]
```diff
@@ -805,10 +805,52 @@ method. It takes 3 arguments: the first is the name of the section as
...
+  # section people_with_block will have headline and
+  # headline_clone elements in it
+  section :people_with_block, People do
+    element :headline_clone, 'h2'
+  end
+end
+```
+
+
+CSS selector can be ommited if you have same locator for all the sections
```
Comment (luke-hill) [inline README.md]:
```
The CSS selector can be omitted if you are re-using the same locator for all
references to the section Class. In order to do this, simply tell SitePrism that
you want to use a default search argument.

<ruby code>
```

EDIT: Delete one of the two lines above.

File: features/section.feature [hunk 11]
```diff
@@ -22,6 +22,11 @@ Feature: Page Sections
     When I navigate to the section experiments page
     Then I can see a section within a section using nested blocks
 
+  Scenario: section within a section using class & blocks
+    When I navigate to the home page
+    Then I can see elements from the parent section
+    Then I can see elements from the block
```
Comment (luke-hill) [inline features/section.feature]: `And`

File: lib/site_prism/element_container.rb [hunk 12]
```diff
@@ -228,14 +228,37 @@ def deduce_iframe_element_find_args(args)
...
-        section_class = args.shift
-      elsif block_given?
-        section_class = Class.new(SitePrism::Section, &block)
+      section_class = deduce_section_class(args, &block)
+      arguments = deduce_search_arguments(section_class, args)
+      [section_class, arguments]
+    end
+
+    def deduce_section_class(args, &block)
+      if args.first.is_a?(Class) && args.first.ancestors.include?(SitePrism::Section)
```
Comment (luke-hill) [inline lib/site_prism/element_container.rb]: Apologies if I'm missing something here, but this condition isn't always successful, even if a block is given.

This means that `base_class` can sometimes be an undefined local variable right?

File: lib/site_prism/element_container.rb [hunk 13]
```diff
@@ -228,14 +228,37 @@ def deduce_iframe_element_find_args(args)
...
+
+      unless klass
+        raise ArgumentError, "You should provide section class either \
+as a block, or as the second argument."
+      end
+      klass
+    end
+
+    def deduce_search_arguments(section_class, args)
+      if args && !args.empty?
```
Comment (luke-hill) [inline lib/site_prism/element_container.rb]: for both this call and the one below we've got a double chain just checking for their presence safety before evaluating their length.

Could we not maybe create some sort of private method like `present?` like I've done in another PR that takes an array input. Or is that over-analysing do you think?

File: lib/site_prism/element_container.rb (hunk 12 above)
Comment (ineverov) [reply lib/site_prism/element_container.rb]: yes, you are right, in this case, the interpreter will treat it as `nil` [on line 242](https://github.com/natritmeyer/site_prism/pull/250/files#diff-0142fe7beaa583d4ce8030ab2a152817R242) and it will use SitePrism::Section as a base class

File: README.md (hunk 7 above)
Comment (ineverov) [reply README.md]: Sorry, English is not my native language, so I appreciate any suggestions how to improve wording

File: lib/site_prism/element_container.rb [hunk 14]
```diff
@@ -228,14 +228,36 @@ def deduce_iframe_element_find_args(args)
...
+    end
+
+    def deduce_search_arguments(section_class, args)
+      extract_search_arguments(args) ||
+        extract_search_arguments(section_class.default_search_arguments) ||
+        raise(ArgumentError, "You should provide search arguments \
+in section creation or set_default_search_arguments within section class")
+    end
+
+    def extract_search_arguments(args)
```
Comment (ineverov) [inline lib/site_prism/element_container.rb]: I guess, this method can be used in the future to do more work, i.e. converting site_prism's search arguments to capybara's (Now they are the same match, but who knows what they'll change in the future). IMO, it's not good to have dependencies on internal implementation (capybara dsl).

File: lib/site_prism/element_container.rb (hunk 12 above)
Comment (luke-hill) [reply lib/site_prism/element_container.rb]: Qu) When I do `var1 || 'string'` inside a ruby interpreter it crashes. Are you sure this will work? Aren't only instance variables `nil` by default? Or am I missing a further point?

File: spec/section_spec.rb [hunk 15]
```diff
@@ -41,6 +41,57 @@ class PageWithAnonymousSection < SitePrism::Page
           .with_message(error_message)
       end
     end
+
+    context 'default search arguments' do
+      class PageWithSectionWithDefaultSearchArguments < SitePrism::Page
```
Comment (luke-hill) [inline spec/section_spec.rb]: I've not fully read through these yet. But this format for tests I prefer a lot! Top work!

File: lib/site_prism/element_container.rb (hunk 12 above)
Comment (ineverov) [reply lib/site_prism/element_container.rb]: Sorry, I told you part of the truth...
```ruby
var1 || 'string' # Raises NameError
```

```ruby
if false # never can be reached
  var1 = 1
end

var1 || 'string' # Works. It's a part of how the ruby interpreter works. =(
```

File: spec/section_spec.rb (hunk 15 above)
Comment (ineverov) [reply spec/section_spec.rb]: 👍

**Comment by luke-hill** [comment]
pin @tgaff @ricmatsui 

I need to have one final read through of this, but I've asked a few questions. If you guys have any comments can you add them in the next 7-10days?

File: spec/section_spec.rb [hunk 16]
```diff
@@ -41,6 +41,57 @@ class PageWithAnonymousSection < SitePrism::Page
...
+        end
+      end
+
+      context 'if default search arguments are not set and no search arguments provided' do
+        it 'should raise ArgumentError' do
+          expect do
+            class ErroredPage < SitePrism::Page
+              section :section, Section
+            end
+          end.to raise_error(ArgumentError)
```
Comment (luke-hill) [inline spec/section_spec.rb]: can we pass a partial / full message into this.

File: spec/section_spec.rb [hunk 17]
```diff
@@ -49,11 +100,43 @@ class PageWithAnonymousSection < SitePrism::Page
...
+      expect(Section.default_search_arguments).to be_nil
+    end
+
+    it { expect(Section).to respond_to(:set_default_search_arguments) }
+
+    it 'should return default search arguments' do
+      expect(BaseSection.default_search_arguments).to eql([:css, '.default'])
+    end
+
+    it 'should return only this section default if they are set' do
```
Comment (luke-hill) [inline spec/section_spec.rb]: I'm confused by this name and what this test is doing different. If it's just testing the fact css / xpath are different can we use shared contexts?

File: spec/section_spec.rb [hunk 18]
```diff
@@ -49,11 +100,43 @@ class PageWithAnonymousSection < SitePrism::Page
...
+      expect(ChildSection.default_search_arguments).to eql([:xpath, '//html'])
+    end
+
+    it 'should parent section default arguments if defaults are not set' do
```
Comment (luke-hill) [inline spec/section_spec.rb]: returns default search arguments of the parent section if defaults are not set

File: spec/sections_spec.rb [hunk 19]
```diff
@@ -22,6 +28,24 @@ class Page < SitePrism::Page
     it { is_expected.to respond_to(:has_single_section?) }
   end
 
+  describe '.section with class and block' do
+    before do
+      expect(subject).to receive(:find_first).and_return(:element)
```
Comment (luke-hill) [inline spec/sections_spec.rb]: should this be allow?

File: spec/sections_spec.rb [hunk 20]
```diff
@@ -22,6 +28,24 @@ class Page < SitePrism::Page
...
+    end
+
+    it 'should be an instance of provided section class' do
```
Comment (luke-hill) [inline spec/sections_spec.rb]: I'm confused by what this test is verifying?

File: README.md [hunk 21]
```diff
@@ -804,12 +804,44 @@ SitePrism allows adding sections to sections) is to call the `section`
...
+end
+```
+
+The CSS selector can be omitted if you are re-using the same locator for all
+references to the section Class. In order to do this, simply tell SitePrism that
+you want to use a default search argument.
+
+```ruby
+class People < SitePrism::Section
+  set_default_serach_arguments '.people'
```
[the remark this answers is not in the document]
Comment (ineverov) [reply README.md]: Oops, sorry.

File: README.md [hunk 22]
```diff
@@ -804,12 +804,44 @@ SitePrism allows adding sections to sections) is to call the `section`
...
+class HomePage < SitePrism::Page
+  # section people_with_block will have headline and
+  # headline_clone elements in it
+  section :people_with_block, People do
+    element :headline_clone, 'h2'
+  end
+end
+```
+
+The CSS selector can be omitted if you are re-using the same locator for all
```
[the remark this answers is not in the document]
Comment (ineverov) [reply README.md]: 👍

File: README.md [hunk 23]
```diff
@@ -804,12 +804,44 @@ SitePrism allows adding sections to sections) is to call the `section`
 method. It takes 3 arguments: the first is the name of the section as
 referred to on the page (sections that appear on multiple pages can be
 named differently). The second argument is the class of which an
-instance will be created to represent the page section, and the third
-argument is a css selector that identifies the root node of the section
+instance will be created to represent the page section, and the following
+arguments are [Capybara::Node::Finders](https://www.rubydoc.info/github/teamcapybara/capybara/master/Capybara/Node/Finders)
+arguments that identifies the root node of the section
```
[the remark this answers is not in the document]
Comment (ineverov) [reply README.md]: 👍

File: spec/section_spec.rb (hunk 16 above)
Comment (ineverov) [reply spec/section_spec.rb]: Will do

File: spec/sections_spec.rb (hunk 19 above)
Comment (ineverov) [reply spec/sections_spec.rb]: allow will not make assertion method is called

File: spec/sections_spec.rb (hunk 20 above)
Comment (ineverov) [reply spec/sections_spec.rb]: So, it is creating an anonymous class [here](https://github.com/natritmeyer/site_prism/pull/250/files#diff-0142fe7beaa583d4ce8030ab2a152817R242) with parent SingleSection and I want to check it

**Comment by luke-hill** [comment]
Merged. Thankyou for your contribution
